meilisearch-core/src/facets.rs

```diff
@@ -0,0 +1,342 @@
+
+impl Deref for FacetFilter {
+    type Target = Vec<Either<Vec<FacetKey>, FacetKey>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl FacetFilter {
+    pub fn from_str(
+        s: &str,
+        schema: &Schema,
+        attributes_for_faceting: &[FieldId],
+    ) -> Result<Self, FacetError> {
+
+        let parsed = serde_json::from_str::<Value>(s).map_err(|e| FacetError::ParsingError(e.to_string()))?;
+        let mut filter = Vec::new();
+        match parsed {
+            Value::Array(and_exprs) => {
+                if and_exprs.is_empty() {
+                    return Err(FacetError::EmptyArray);
+                }
+                for expr in and_exprs {
+                    match expr {
+                        Value::String(s) => filter.push(Either::Right(FacetKey::from_str(
+                            &s,
+                            schema,
+                            attributes_for_faceting,
+                        )?)),
+                        Value::Array(or_exprs) => {
+                            if or_exprs.is_empty() {
+                                return Err(FacetError::EmptyArray);
+                            }
+                            let mut inner = Vec::new();
+                            for expr in or_exprs {
+                                match expr {
+                                    Value::String(s) => inner.push(FacetKey::from_str(
+                                        &s,
+                                        schema,
+                                        attributes_for_faceting,
+                                    )?),
+                                    bad_value => {
+                                        return Err(FacetError::UnexpectedToken { expected: &["String"], found: bad_value.to_string() })
+                                    }
+                                }
+                            }
+                            filter.push(Either::Left(inner));
+                        }
+                        bad_value => return Err(FacetError::UnexpectedToken { expected: &["String", "Array"], found: bad_value.to_string() }),
+                    }
+                }
+                return Ok(Self(filter));
+            }
+            bad_value => Err(FacetError::UnexpectedToken { expected: &["Array"], found: bad_value.to_string() }),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+#[repr(C)]
+pub struct FacetKey(FieldId, String);
+
+impl FacetKey {
+    pub fn new(field_id: FieldId, value: String) -> Self {
+        let value = match value.cow_to_lowercase() {
+                Cow::Borrowed(_) => value,
+                Cow::Owned(s) => s,
+        };
+        Self(field_id, value)
+    }
+
+    fn from_str(
+        s: &str,
+        schema: &Schema,
+        attributes_for_faceting: &[FieldId],
+    ) -> Result<Self, FacetError> {
+        let mut split = s.splitn(2, ":");
```

Comment:
```suggestion
        let mut split = s.splitn(2, ':');
```

Reply:
I think we should better specify `splitn(3, _)` here and check if the 3rd doesn't exists `split.next().is_none()`.
This way we can detect syntax error like `["colors:blue:green"]`.

Reply:
There also is a little issue that we should note somewhere: what if the user write this `["colors:\"top:blue\""]`?
We will split on the second `:` and we must not do that.

Therefore you should not `splitn(3)` like I proposed in my comment aboce, but we should implement a better parser (using Pest) in a next PR, later, later.

Reply:
yes I thought of that, and there is this other case that's breakign too: `['\"foo:bar\":hello]` that I don;t take into account at all